### PR TITLE
Rework docker usage in LocalStack to make use of the docker client

### PR DESCRIPTION
**Pull request by dfangl**
This PR replaces run() calls throughout the code base with calls of the docker client, to reduce possible vulnerability to command injection.

To prevent issues with circular imports. the run() method and FuncThread() class had to be moved from bootstrap.py to a separate package.

### CODE REVIEW

File: localstack/services/awslambda/lambda_executors.py [hunk 1]
```diff
@@ -13,6 +13,7 @@
 from multiprocessing import Process, Queue
 from typing import Dict, Tuple, Union
 
+import localstack.utils.docker
```
Comment (whummer) [inline localstack/services/awslambda/lambda_executors.py:16]: nitpick: So far, we've usually used relative (not absolute) import names. We could continue using this pattern, e.g. something like: 
```
from localstack.utils import docker as docker_utils
```
(not critical at all, though - shouldn't hold back the merge)

File: localstack/services/install.py [hunk 2]
```diff
@@ -296,20 +298,24 @@ def install_stepfunctions_local():
...
-                dc=config.DOCKER_CMD, dn=docker_name, tgt=INSTALL_DIR_INFRA
-            )
+        DOCKER_CLIENT.copy_from_container(
+            docker_name, local_path=INSTALL_DIR_INFRA, container_path="/home/stepfunctionslocal/"
         )
-        run("mv %s/stepfunctionslocal/*.jar %s" % (INSTALL_DIR_INFRA, INSTALL_DIR_STEPFUNCTIONS))
+
+        path = Path(f"{INSTALL_DIR_INFRA}/stepfunctionslocal/")
+        for file in path.glob("*.jar"):
+            file.rename(f"{INSTALL_DIR_STEPFUNCTIONS}/{file.name}")
```
Comment (whummer) [inline localstack/services/install.py:318]: nit: I'm guessing `file.rename` can interpret `"/"` as path delimiter on all systems, but to be on the safe side, we could do:

```
file.rename(os.path.join(INSTALL_DIR_STEPFUNCTIONS, file.name))
```
or:
```
file.rename(Path(INSTALL_DIR_STEPFUNCTIONS) / file.name)
```

File: localstack/utils/bootstrap.py [hunk 3]
```diff
@@ -236,68 +223,56 @@ def get_docker_image_details(image_name=None):
...
         # try to extract from existing running container
         container_name = get_main_container_name()
-        version = run(
-            "%s exec -it %s bin/localstack --version" % (docker_cmd, container_name),
-            print_error=False,
+        version, _ = DOCKER_CLIENT.exec_in_container(
+            container_name, interactive=True, command=["bin/localstack", "--version"]
         )
-        version = version.strip().split("\n")[-1]
+        version = version.decode(config.DEFAULT_ENCODING).strip().splitlines()[-1]
```
Comment (whummer) [inline localstack/utils/bootstrap.py:260]: nit: we could use the `to_str(..)` util function here:
```
         version = to_str(version).strip().splitlines()[-1]
```
(also on line 273 below)

File: localstack/utils/docker.py [hunk 4]
```diff
@@ -163,85 +263,131 @@ def copy_into_container(
...
+    def get_container_logs(self, container_name_or_id: str, safe=False) -> str:
+        """Get all logs of a given container"""
+        cmd = self._docker_cmd()
+        cmd += ["logs", container_name_or_id]
+        try:
+            return safe_run(cmd)
         except subprocess.CalledProcessError as e:
-            if "No such image" in e.stdout.decode(config.DEFAULT_ENCODING):
-                raise NoSuchImage(docker_image, stdout=e.stdout, stderr=e.stderr)
+            if not safe:
```
Comment (whummer) [inline localstack/utils/docker.py:311]: I see that you've copied the code from above, but wondering if we had this wrong to begin with. I guess we should change this to:
```
if safe:
    return ""
```
?

**Review by whummer — APPROVED** [review]
Looks awesome, thanks @dfangl ! 🚀 Excited to get this merged - super important changes. 👍

File: localstack/services/awslambda/lambda_executors.py (hunk 1 above)
Comment (dfangl) [reply localstack/services/awslambda/lambda_executors.py:16]: Thanks, did not catch that one, pycharm seems to had a hiccup there when moving PortMappings from bootstrap to docker.py

**Comment by whummer** [comment]
Hi @colin-ife-snyk , thanks for reaching out. Will get back to you directly via email.. Thanks

**Pull request merged**